codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/testutil/EventStreamTestTools.kt

```diff
@@ -0,0 +1,173 @@
+        val generator = requirements.renderGenerator(codegenContext, test, protocol)
+
+        test.project.lib {
+            when (variety) {
+                EventStreamTestVariety.Marshall -> writeMarshallTestCases(testCase, generator)
+                EventStreamTestVariety.Unmarshall -> writeUnmarshallTestCases(testCase, codegenTarget, generator)
+            }
+        }
+        test.project.compileAndTest()
+    }
+
+    private fun <C : CodegenContext> generateTestProject(
+        requirements: EventStreamTestRequirements<C>,
+        codegenContext: C,
+        codegenTarget: CodegenTarget,
+    ): TestEventStreamProject {
+        val model = codegenContext.model
+        val symbolProvider = codegenContext.symbolProvider
+        val operationShape = model.expectShape(ShapeId.from("test#TestStreamOp")) as OperationShape
+        val unionShape = model.expectShape(ShapeId.from("test#TestStream")) as UnionShape
+
+        val project = TestWorkspace.testProject(symbolProvider)
+        val operationSymbol = symbolProvider.toSymbol(operationShape)
+        project.withModule(ErrorsModule) {
+            val errors = model.shapes()
+                .filter { shape -> shape.isStructureShape && shape.hasTrait<ErrorTrait>() }
+                .map { it.asStructureShape().get() }
+                .toList()
+            when (codegenTarget) {
+                CodegenTarget.CLIENT -> CombinedErrorGenerator(model, symbolProvider, operationSymbol, errors).render(this)
+                CodegenTarget.SERVER -> ServerCombinedErrorGenerator(model, symbolProvider, operationSymbol, errors).render(this)
+            }
+            for (shape in model.shapes().filter { shape -> shape.isStructureShape && shape.hasTrait<ErrorTrait>() }) {
+                StructureGenerator(model, symbolProvider, this, shape as StructureShape).render(codegenTarget)
+                requirements.renderBuilderForShape(this, codegenContext, shape)
+            }
+        }
+        project.withModule(ModelsModule) {
+            val inputOutput = model.lookup<StructureShape>("test#TestStreamInputOutput")
+            recursivelyGenerateModels(model, symbolProvider, inputOutput, this, codegenTarget)
+        }
+        project.withModule(RustModule.Output) {
+            operationShape.outputShape(model).renderWithModelBuilder(model, symbolProvider, this)
+        }
+        return TestEventStreamProject(
+            model,
+            codegenContext.serviceShape,
+            operationShape,
+            unionShape,
+            symbolProvider,
+            project,
+        )
+    }
+
+    private fun recursivelyGenerateModels(
+        model: Model,
+        symbolProvider: RustSymbolProvider,
+        shape: Shape,
+        writer: RustWriter,
+        mode: CodegenTarget,
+    ) {
+        for (member in shape.members()) {
+            val target = model.expectShape(member.target)
+            if (target is StructureShape || target is UnionShape) {
```

Comment:
Is the reason why we only render these shape types that we're not testing with more complex models? If yes, I would crash here upon encountering an unrendered shape type, lest a future dev be puzzled when adding more shapes to the event stream tests.

Reply:
Good call. Fixed in c608a775bf02ce33b73f6827bb12b6a8b94ad6ed.

Reply:
Actually, I forgot to remove the `if (target is StructureShape || target is UnionShape)` while testing 🤦‍♂️ 

Correctly fixed in 38598f2e318b201d78a81d8e2c1185587f962c4e.